### ISSUE TITLE
[BUG] Models do not use cuDNN RNN

### ISSUE THREAD

**Issue by dataframing** [BrikerMan/Kashgari#125]
## Checklist

- [x] I have searched in [existing issues](https://github.com/BrikerMan/Kashgari/issues?utf8=%E2%9C%93&q=is%3Aissue+) but did not find the same one.

## Environment

* OS [e.g. Mac OS, Linux]: macOS Mojave (10.14.5)
* requirements.txt: **Omitted since this is a within-package issue. See below.**

## Issue Description

None of the concrete `SequenceLabelingModel` sub-classes are able to make use of their cuDNN-aware RNN layers.

### What

Hi there,

First of all, thank you for this repository! It's well-organized. That said, I was trying to understand how the underlying models work and noticed #31 (specifically, your comment [here](https://github.com/BrikerMan/Kashgari/issues/31#issuecomment-469595699)) where you mentioned that all of the models can use the main RNN layers' cuDNN-aware analogues via

```python
import kashgari
kashgari.config.use_CuDNN_cell = True
```

I had an issue with this before, so I tried what you said and the model fit! I wanted to understand how the library does it, but on further inspection I noticed that this wasn't quite true. See below, where for the first example we explicitly set it to `False`, build the model, and upon inspection the post-BERT recurrent layer is truly an `LSTM` (correct). Then, in a fresh environment, we do the same example but set `True`, but the layer is not an instance of `CuDNNLSTM` — it's still an `LSTM`.

```python
import kashgari
kashgari.config.use_CuDNN_cell = False # <<<

from kashgari.corpus import CoNLL2003Corpus
from kashgari.embeddings import BERTEmbedding
from kashgari.tasks.seq_labeling import BLSTMModel


# Load our data and build our model.
x_data, y_data = CoNLL2003Corpus.get_sequence_tagging_data()
bert = BERTEmbedding(name_or_path='bert-base-cased')
model = BLSTMModel(embedding=bert)
model.build_model(x_train=x_data, y_train=y_data)

# This assertion passes.
from keras.layers import LSTM
assert model.model.layers[-4].layer.__class__.__name__ == LSTM.__name__
```

```python
import kashgari
kashgari.config.use_CuDNN_cell = True # <<<

from kashgari.corpus import CoNLL2003Corpus
from kashgari.embeddings import BERTEmbedding
from kashgari.tasks.seq_labeling import BLSTMModel


# Load our data and build our model.
x_data, y_data = CoNLL2003Corpus.get_sequence_tagging_data()
bert = BERTEmbedding(name_or_path='bert-base-cased')
model = BLSTMModel(embedding=bert)
model.build_model(x_train=x_data, y_train=y_data)

# This raises an error.
from keras.layers import CuDNNLSTM
assert model.model.layers[-4].layer.__class__.__name__ == CuDNNLSTM.__name__
```

For the sequence labeling models, the first issue is that they don't import the `LSTMLayer` alias from `kashgari.layers`, but instead import from and directly use `keras.layers.LSTM`. However, the classification models _do_ properly import `GRULayer` and `LSTMLayer` from `kashgari.layers` but they still don't use the cuDNN-aware RNN layers:

```python
import kashgari
kashgari.config.use_CuDNN_cell = True # <<<

from kashgari.corpus import CoNLL2003Corpus
from kashgari.tasks.classification.models import BLSTMModel
from kashgari.corpus import SMP2017ECDTClassificationCorpus

x_data, y_data = SMP2017ECDTClassificationCorpus.get_classification_data()
model = BLSTMModel()
model.build_model(x_train=x_data, y_train=y_data)

# This assertion fails.
from keras.layers import CuDNNLSTM
assert model.model.layers[-2].layer.__class__.__name__ == CuDNNLSTM.__name__
```

So the issue must be with how we're defining `LSTMLayer` in `kashgari.layers`.

### Reproduce

I think the issue can be minimally isolated to the following example:

```python
import kashgari
kashgari.config.use_CuDNN_cell = True

from kashgari.layers import LSTMLayer
from keras.layers import CuDNNLSTM

# This assertion fails.
assert LSTMLayer.__name__ == CuDNNLSTM.__name__
```

### Other Comment

As a suggestion/solution, I copied most of the code from the sequence-labeling `BLSTMModel` [source](https://github.com/BrikerMan/Kashgari/blob/master/kashgari/tasks/seq_labeling/blstm_model.py), but avoided using the nested `if/else` for defining the `LSTMLayer` alias in `kashgari.layers` and instead defined it at the module level, but as a conditional assignment.

```python
import kashgari
kashgari.config.use_CuDNN_cell = True # <<<

from keras.layers import Bidirectional, LSTM, CuDNNLSTM
from keras.layers import Dense, Dropout, TimeDistributed, Activation
from keras.models import Model

from kashgari.corpus import CoNLL2003Corpus
from kashgari.embeddings import BERTEmbedding
from kashgari.tasks.seq_labeling.base_model import SequenceLabelingModel


LSTMLayer = CuDNNLSTM if kashgari.config.use_CuDNN_cell else LSTM


class CudnnAwareBLSTMModel(SequenceLabelingModel):
    __architect_name__ = 'BLSTMModel'
    __base_hyper_parameters__ = {
        'lstm_layer': {
            'units': 256,
            'return_sequences': True
        }, 'dropout_layer': {
            'rate': 0.4
        }
    }

    def _prepare_model(self):
        embed_model = self.embedding.model

        # Note: we're using `LSTMLayer`.
        blstm_layer = Bidirectional(LSTMLayer(**self.hyper_parameters['lstm_layer']))(embed_model.output)
        dropout_layer = Dropout(**self.hyper_parameters['dropout_layer'])(blstm_layer)
        time_distributed_layer = TimeDistributed(Dense(len(self.label2idx)))(dropout_layer)
        activation = Activation('softmax')(time_distributed_layer)

        self.model = Model(embed_model.inputs, activation)

    def _compile_model(self):
        loss_f = 'categorical_crossentropy'
        optimizer = 'adam'
        metrics = ['accuracy']

        self.model.compile(loss=loss_f, optimizer=optimizer, metrics=metrics)


# Build our CuDNN-aware model.
x_data, y_data = CoNLL2003Corpus.get_sequence_tagging_data()
bert = BERTEmbedding(name_or_path='bert-base-cased')
model = CudnnAwareBLSTMModel(embedding=bert)

# Note: I can't compile the model on my local, non-GPU machine. But the idea holds.
# model.build_model(x_train=x_data, y_train=y_data)
model._prepare_model()

# This works!
assert model.model.layers[-4].layer.__class__.__name__ == CuDNNLSTM.__name__
```

Implementation-wise, I think it makes more sense to avoid importing a `config` singleton and instead have a `use_cudnn_rnn` flag on the `BaseModel` instance that:

1. Defaults to `False`.
2. If `True`, checks for available GPUs and tensorflow backend.
3. If successful, uses `CuDNN[LSTM/GRU]` instead of aliasing.

Please let me know if I'm wrong and/or unclear! If it's fine with you, I'm happy to open a PR with a regression test specifically for this issue. Thank you.

**Comment by BrikerMan** [comment]
Thanks for the detailed issue. This is a legacy issue, I might won’t fix it in this version. But we have already fully rewritten Kashgari in tf.keras and it almost finished. It will use CuDNNlSTM and CuDNNGRU in labeling model. PS: Which reminded me to check the RNN layer.  

The tk.keras version written more clean, well organized and have better performance. We will focus on this version afterwards. Maybe you could try it out. 

tf.keras version introduces several breaking changes. Remember to checkout the document.

**Comment by alexwwang** [comment]
To put it simple, as you mentioned you use a MacOS, I assume you use a
Macbook (pro), which normally doesn't have an Nvidia display card
supporting CuDA/CuDNN in it. Am I right?

So the fact is, only if you have CuDA/CuDNN installed and back-support
hardware installed, the CuDNNLSTM layer could work properly, you may refer
to Keras' documents to clarify this fact.

Then the mechanism in kashgari is, there are lines of code detecting if
your machine has a Nvidia CuDA card properly installed. If yes, the
kashgari.layers.LSTM would be set to CuDA one automatically or normal lstm
one forever.
The same to RNN ones.

So you may check if this is the situation you met with.

**Comment by alexwwang** [comment]
And sorry for making you confused by the master branch.

There were actually two lines to develop classification and labeling tasks
respectively. So each line has its own refactored codes or features that
have not been updated to another yet.

But recently, a new refactored branch `tf-keras.version` is near ready to
release, which is well defined on both tasks and you may check if it solves
your problem.

The plan is, the original master branch would be frozen while we will focus
on build and improve the new branch based on TF.keras. So you're always
warmly welcome to open a PR based on this new branch if you have any idea
to improve it. What's more, if you would like to improve the original one,
personally, I am also happy to see that.

Angain thank you for your advice and look forward to  your feedback.

**Comment by BrikerMan** [comment]
@dataframing Thanks again for the detailed issue. Fixed the issue in master(legacy) branch. From now on, we will focus on tf.keras version, and it will become the master branch after the first release. It's a  totally rewritten version, with cleaner code and powerful features. Remember to check it out. 

This is how we did this in tf.keras version, it will use CuDNN cell by default.
https://github.com/BrikerMan/Kashgari/blob/3faaf8ccb823bb8cf1cc2e2ffb7ad2fac3ac6a7a/kashgari/layers/__init__.py#L19
